__tests__/produce.ts

```diff
@@ -274,3 +297,68 @@ it("works with generic parameters", () => {
     let arr: ReadonlyArray<typeof val> = 0 as any
     insert(arr, 0, val)
 })
+
```

Comment:
In these two test cases I tried to capture two real-life scenarios:
1. Your generic state shape is defined as mutable
2. The generic state shape is defined as immutable

Given those two scenarios (tests), the tests below verify a few things:
1. draft is always mutable, regardless the base state
2. if `state: State`, then ideally `produce(state, fn)` and `produce(fn)(state)` should result in `State` again

Reply:
Ok, I think in the latest push I've found the solution that satisfies all those constraints! In the latest, the curried producer has it's own generic argument, and basically, what goes in, goes out.  That means: immutable types in, immutable types out. Mutable types in, mutable types out. However, the generic argument has to at least satisfy that it is at least the `Immutable<T>` where `T` is the first argument of the recipe. 

I think this solves all cases quite neatly. The only downside is that the tests that use `exactType` on the curried function now fails, because the curried producer has a wider type than the recipe. But technically that is actually valid, as the following statement is correct:

`produce((draft: { count: 0 }) => { draft.count++ }) ({ count: 7, messages: [] })`

The above now neatly does produce a new state of type `{ count, messages }` rather than just `{ count } ` which it would be, based on the recipe

Reply:
cc @aleclarson